### DISCUSSION TITLE
How to get Remote Address with httpx?

### DISCUSSION THREAD

**Discussion by baiyyee** [encode/httpx#1569] [Q&A] [answered]
With **requests**, we used ```response.raw._original_response.fp.raw._sock.getpeername()[0]``` to get **Remote Address**, but it's not work with **httpx**, so how to get Remote Address with httpx?

![image](https://user-images.githubusercontent.com/4414171/114420694-1eb71e00-9be7-11eb-85b5-2d5eea6ae904.png)

**Comment by lovelydinosaur** [comment]
Ah great question!

Nope, we don't support that at the moment.

Notes to the team for how we'd need to address this...

1. Interface-wise this requires us adding something like a `remote_addr` extension to the Transport API.
2. Implementation-wise our [HTTPCore backends](https://github.com/encode/httpcore/tree/master/httpcore/_backends) would need to expose this information. (As API on the socket stream classes)
3. This would be available initially via `response.extensions["remote_addr"]`, and perhaps also at a later date via a dedicated `response.remote_addr` attribute.

Other thoughts here...

* Presumably we might also want to support `local_addr`, based on [`socket.getsockname()`](https://docs.python.org/3/library/socket.html#socket.socket.getsockname).
* We'd need to think through exactly what format this has. Two-tuple of `(bytes, int)`?
* What behaviour are we expecting when used with Unix Domain Sockets?
* Presumably the extension key would be absent from WSGI/ASGI transports.

Note that client/server addresses also came up briefly when discussing server-side usages... https://github.com/encode/httpx/discussions/1527#discussioncomment-517519

Looking at the ASGI spec is helpful here too... https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope

Thinking about ensuring consistency with being able to provide this on either requests or responses impacts naming here.

Having typed this through it seems to me that a good naming here would be `server_addr` and `client_addr`, which is preferable to local/remote because it is an absolute naming system, rather than a relative one. (Ie which side is "local" depends on who's talking.)

**Comment by SK-415** [comment, marked as answer]
Message to future people looking for this feature.

It's now support with `httpx>=0.21.0` or technically `httpcore>=0.14.0`. See [request extensions](https://www.encode.io/httpcore/extensions/#extra-network-information).

```py
import httpx

with httpx.Client() as client:
    r = client.get("https://example.com")
    print(r.extensions["network_stream"].get_extra_info("server_addr"))
    print(r.extensions["network_stream"].get_extra_info("client_addr"))
```
